eeps/eep-0079.md

```diff
@@ -0,0 +1,784 @@
+
+An update operation fails with a `{badfield,FN}` error if:
+
+* The native-record update expression references the field FN which is
+  not defined (in the structure definition).
+
+Native-record update expressions are not allowed in guards.
+
+### Anonymous update of native records
+
+The following syntax allows update any record having the given fields:
+
+```erlang
+Expr#_{Field1=Expr1, ..., FieldN=ExprN}
+```
+
+An anonymous update operation fails with a `{badfield,FN}` error if:
+
+* The native-record update expression references the field FN which is
+  not defined (in the structure definition).
+
+### Pattern matching over native records
+
+A pattern that matches a certain native-record value is created in the
+same way as a native-record is created.
+
+The syntax:
+
+```erlang
+#Name{Field1 = Expr1, ..., FieldN = ExprN}
+#Module:Name{Field1 = Expr1, ..., FieldN = ExprN}
+```
+
+Here, `Expr1`, .. `ExprN` are patterns, and Field names must be atoms.
+
+Pattern matching fails if:
+
+* The native-record value is not visible at the call site, that is,
+  an attempt is made to match out a a value from a non-exported record
+  from a module other than its defining module.
+
+* The pattern references a FieldK and the native-record value does not
+  contain this field.
+
+Note, however, that it is possible to match on the name of a
+non-exported record. Thus, if the `match_name/1` function in the
+following example is called with an instance of record `r` defined in
+`some_module`, it will succeed even if the record is not exported:
+
+```erlang
+-module(example).
+-export([match_name/1]).
+
+match_name(#some_module:r{}) ->
+  ok.
+```
+
+<!-- -->
+
+```erlang
+-module(some_module).
+-export([get_r/1]).
+
+-record #r{a}.
+
+get_r(N) -> #r{a=N}.
+```
+
+The following syntax allows matching any record having the named fields:
+
+```erlang
+#_{Field1 = Expr1, ..., FieldN = ExprN}
+```
+
+### Checking whether a record is current
+
+Do we need a way to check that the native-record is the current version?
+
+Yes, we should add a BIF that essentially does the following:
+
+```erlang
+record:get_fields(Record) =:= record:get_fields(#Module:Record{})
+```
+
+but guaranteed to always work and be more efficient.
+
+TODO: What should the name of the BIF be?
+
+We should also have a BIF that checks whether an instance of a native
+record referes to the current definition of the native record.
+
+TODO: What should the name of that BIF be?
+
+### Fetching field index
+
+Fetching the record index using the `Rec#name.field` syntax is not
+supported, because there is no way it can actually be used, since neither
+ETS nor `element/2` will work with native records.
+
+### Native-record guard BIFs
+
+#### `is_record/2`
+
+We define a pseudo-BIF `is_record/2`, which will be translated by
+the compiler to instructions to test whether `Term` is a native
+record.
+
+```erlang
+is_record(Term :: dynamic(), #Module:Name) -> boolean().
```

Comment:
I really don't like the idea of allowing free `#r` or `#m:r` as in `is_record(X, #r)`, because those fragments do not have any semantic meaning on their own, but they will have to be handled as something that may occur in any expression position. I think the rationale further below is wrong in this case: it's much worse to introduce pseudo-expressions that have no defined meaning outside a specific context. And you won't be using that syntax for the API functions `records:create(Module, Name, ....)` etc. anyway.